new connections and variables doc (#1877)
astronomer/docs: astro/manage-connections-variables.md

---

> sidebar_label: 'Manage Airflow Connections and Variables'

Reviewer: ```suggestion
sidebar_label: 'Manage Airflow connections and variables'
```

> title: 'Manage Airflow Connections and Variables'

Reviewer: ```suggestion
title: 'Manage Airflow connections and variables'
```

id: manage-connections-variables
description: "Manage Airflow Connections and Variables"
---

> ## **Introduction**

Reviewer: ```suggestion

```

### Create Objects using the Airflow API

With Astro, you have access to your [Deployment’s Airflow API](https://docs.astronomer.io/astro/airflow-api). You can use this API to create and export Airflow Connections and Variables. Airflow stores these objects in the Airflow metadata database in the dataplane. Teams setting up large deployments with many Airflow objects may want to create them with the Airflow API. You can also use the API to export Airflow objects, but the API response will omit secret values from an export.

### Create Objects using the Airflow CLI

Unlike local development, you cannot use the [Airflow CLI](https://docs.astronomer.io/astro/test-and-troubleshoot-locally#run-airflow-cli-commands) with an Astro Deployment. If you’d like to create multiple objects at once, you can use the Airflow API, environment variables, or a [secretes backend](https://docs.astronomer.io/astro/secrets-backend).

### Environment Variables on Astro

> As mentioned earlier, you can create Airflow Objects with environment variables. On Astro, you can create Airflow Objects through Astro UI and CLI. Do create them with the UI, you'd go to the [deployment](https://docs.astronomer.io/astro/configure-deployment-resources) page.

Reviewer: ```suggestion
As mentioned earlier, you can create Airflow Objects with environment variables. On Astro, you can create Airflow Objects through Astro UI and CLI. To create them with the UI, you'd go to the [deployment](https://docs.astronomer.io/astro/configure-deployment-resources) page.
```

You can also create environment variables for your Astro Deployments with the Astro CLI. For example to load environment variables from a `.env` file run:

```bash
astro deployment variable create --deployment-name <deployment-name> --load --env <env-file>
```

### Secret Backend

> You can store your Airflow Connections and Variables as Environment Variables in a [secrets backend](https://docs.astronomer.io/astro/secrets-backend). A secrets backend is a centralized location that complies with your organization's security requirements. Some common secret backends that Astro integrates with are AWS [Secretes Manager](https://aws.amazon.com/secrets-manager/), [Hashicorp Vault](https://www.vaultproject.io/), [Google Cloud Secret Manager](https://cloud.google.com/secret-manager), and [Azure Key Vault](https://azure.microsoft.com/en-us/products/key-vault/). Storing your Airflow objects this way is recommended for production secrets that allow DAGs to access critical systems or information. Secrets backends provide one centralized location for your secret Airflow objects. You can easily share objects between Airflow deployments once you connect each deployment to the secrets backend. One downside of this method is that you will not see your Airflow Connections or Variables in the Airflow UI. The secrets backendare does not store them in the metadata database. You can read our [documentation](https://docs.astronomer.io/astro/secrets-backend) for instructions on integrating your Astro deployments with popular secret backends.

Reviewer: ```suggestion
You can store your Airflow Connections and Variables as Environment Variables in a [secrets backend](https://docs.astronomer.io/astro/secrets-backend). A secrets backend is a centralized location that complies with your organization's security requirements. Some common secret backends that Astro integrates with are AWS [Secrets Manager](https://aws.amazon.com/secrets-manager/), [Hashicorp Vault](https://www.vaultproject.io/), [Google Cloud Secret Manager](https://cloud.google.com/secret-manager), and [Azure Key Vault](https://azure.microsoft.com/en-us/products/key-vault/). Storing your Airflow objects this way is recommended for production secrets that allow DAGs to access critical systems or information. Secrets backends provide one centralized location for your secret Airflow objects. You can easily share objects between Airflow deployments once you connect each deployment to the secrets backend. One downside of this method is that you will not see your Airflow Connections or Variables in the Airflow UI. The secrets backend are does not store them in the metadata database. You can read our [documentation](https://docs.astronomer.io/astro/secrets-backend) for instructions on integrating your Astro deployments with popular secret backends.
```

### How to push Local Connections and Variables to an Astro Deployment

You may want to transfer Connections and Variables created locally to an Astro deployment. For example, once you have a few DAGs working locally, you may want to push them along with needed Airflow objects up to an Astro deployment. You can push Airflow objects from the local Airflow metadata database, settings file, or .env file to a Deployment’s environment variables. The following instructions will walk you through this, starting with the objects in the local database.

First make sure you have your local Airflow environment running with: 

```bash
astro dev start
```

Next make sure all the objects in your settings file are in the local database by running:

```bash
astro dev object import --settings-file <settings-file-location>
```

Export all of your local Airflow connections to an environment file with:

```bash
astro dev object export --env <env-file-location>
```

Now that all your objects are in an environment file you can push them up to an Astronomer Deployment as environment variables with the following command:

```bash

> astro dpeloyment variable create --deployment-name "<deployment name>" --load --env <env-file-location> --secret

Reviewer: ```suggestion
astro deployment variable create --deployment-name "<deployment name>" --load --env <env-file-location> --secret
```

---
sidebar_label: 'Manage Airflow connections and variables'
title: 'Manage Airflow connections and variables on Astro'

> id: airflow-connections-variables

Reviewer 2: ```suggestion
id: manage-connections-variables
```

---

You can store Airflow [connections](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html) and [variables](https://airflow.apache.org/docs/apache-airflow/stable/howto/variable.html) on Astro using several different methods. Each of these methods has advantages and limitations related to their security, ease of use, and performance. Similarly, there are a few strategies for testing connections and variables on your local computer and deploying these configurations to Astro.

Use this document to understand all available options for managing variables and connections in both local Astro projects and Astro Deployments. 

To set system-level environment variables to [change Airflow configuration settings](https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html), see [Set environment variables](environment-variables.md). To learn more about Airflow variables and connections, see [Manage connections in Apache Airflow](https://docs.astronomer.io/learn/connections)** and [Variables in Apache Airflow: The Guide](https://marclamberti.com/blog/variables-with-apache-airflow/).

## Prerequisites

> - A locally hosted Astro project. See [Create a project](create-project.md).

Reviewer 3: Yes that's exactly what happened - I added a suggestion with the corrected link!

The limitations of this strategy are:

- You can only automatically export your variables to Astro as environment variables
- Managing many connections or variables this way can become unwieldy.
- It's not the most secure option for sensitive variables. 
- You'll lose your connections and variables if you delete your metadata database with `astro dev kill`.

To create Airflow connections and variables in the Airflow UI:

1. Run the following command to start your local Airflow environment:

    ```sh
    astro dev start
    ```

2. Access the Airflow UI at `localhost:8080`.
3. Create your connections and variables. See:

    - [Managing Variables](https://airflow.apache.org/docs/apache-airflow/stable/howto/variable.html)
    - [Create a Connection in the UI](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#creating-a-connection-with-the-ui)

Airflow connections and variables configured in the Airflow UI are stored in your local environment's metadata database. You can view all connections and variables set this way in the Airflow UI under **Admin** > **Connections** and **Admin** > **Variables**. 

#### Deploy to Astro

> To export connections and variables from the Airflow UI to a Deployment on Astro, run:

Author: deploying code is sending it to an environment where it will be run. exporting is just like exporting to a file

```sh
astro dev object export --env-export
astro deployment variable create --deployment-name="My Deployment" --load --env .env
```

These commands format your connections and variables as system environment variables, then push the variables to Astro. Connections are formatted as [URIs](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#uri-format).

### Manage connections and variables locally as environment variables

> You can use Airflow's system-level environment variables to store connections and variables for use by your DAGs. To do this locally, add each environment variable on its own line in your Astro project `.env` file. When you start your Airflow environment, the Astro CLI builds these variables into your Docker image and makes them available to your Airflow components. However, your connections and variables are not available to view in the Airflow UI. 

Reviewer 3: ```suggestion
You can use Airflow's system-level environment variables to store connections and variables for use by your DAGs. To do this locally, add each environment variable in its own line in your Astro project `.env` file. When you start your Airflow environment, the Astro CLI builds these variables into your Docker image and makes them available to your Airflow components. However, your connections and variables are not available to view in the Airflow UI. 
```

The benefits of managing Airflow connections and variables in your `.env` file are: 

- If your local metadata database is corrupted or accidentally deleted, you still have access to all of your connections and variables. 
- You can export the file to an Astro Deployment using the Astro CLI.
- You can manage connections and variables as code, allowing you to move and copy connections and variables between Deployments.

The limitations of this method are:

- Connections can't be viewed or managed from the Airflow UI.
- You have to restart your local Airflow environment whenever you make changes to your `.env` file. 
- You can't store your variables as secrets until you deploy to Astro.
- Connections are more difficult to format as URIs and JSON.

#### Set Airflow variables as local environment variables

To define an Airflow variable as an environment variable, use the following format:

```text
AIRFLOW_VAR_<VAR_NAME>='<VALUE>'
```

> To retrieve the value of an Airflow variable in a DAG, use `os.getenv('AIRFLOW_VAR_<VAR_NAME>','<default-value>')`. This method is reduces the number of Airflow metadata database requests. However, it's less secure. Astronomer does not recommend using `os.getenv` with secret values because calling these values with the function can print them to your logs. Replace `<default_value>` with a default value to use if Airflow can't find the environment variable. Typically, this is the value you defined for the environment variable in the Cloud UI.

Reviewer 3: ```suggestion
To retrieve the value of an Airflow variable in a DAG, use `os.getenv('AIRFLOW_VAR_<VAR_NAME>','<default-value>')`. This method reduces the number of Airflow metadata database requests. However, it's less secure. Astronomer does not recommend using `os.getenv` with secret values because calling these values with the function can print them to your logs. Replace `<default_value>` with a default value to use if Airflow can't find the environment variable. Typically, this is the value you defined for the environment variable in the Cloud UI.
```

#### Set Airflow connections as local environment variables

To define an Airflow connection as an environment variable in Runtime 4.2 and earlier, add the connection as a [URI](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#uri-format) to your `.env` file in the following format:

```text
AIRFLOW_CONN_<CONN_ID>='<connection-uri>'
```

In Astro Runtime version 5.0 and later, you can also use JSON format to store connections. See [JSON format example](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#json-format-example). When using the JSON format in environment variables, the JSON object must be defined in a single, unbroken line. 

> To use a connection in your DAG, you specify the connection ID like you normally would for a connection set in the Airflow UI.

Reviewer 3: Does this need a code example?

:::tip

If you prefer to format your Airflow connections in the Airflow UI but you want to store them as environment variables, run the following command to export all connections  from the metadata database to your `.env` file:

```sh
astro dev object export --connections --env-export
```

:::

#### Deploy to Astro

To push your environment variables to a Deployment, run:

```sh
astro deployment variable create --deployment-name="My Deployment" --load --env .env
```

After running this command, open the Cloud UI and mark any environment variables storing secret values as **Secret**. See [Environment variables](environment-variables.md).

### Manage Airflow connections and variables locally using `airflow_settings.yaml`

You can use the Astro CLI to manage local Airflow connections and variables with an [`airflow_settings.yaml` file](https://docs.astronomer.io/astro/develop-project#configure-airflow_settingsyaml-local-development-only). The CLI automatically adds all Airflow connections, variables, and pools listed in this file to your local metadata database. 

> The benefits of using `airflow_settings.yaml` are:

Reviewer 3: I'm starting to feel like a comparison table of the benefits/drawbacks with these strategies might be useful, but if we're increasing the number of options, that can get a bit unwieldy.

## Choose a strategy for managing connections and variables on Astro

You can manage connections and variables on Astro similarly to how you manage them locally. However, some features available on your local deployments may not be available on Astronomer.

Regardless of the strategy you choose, keep in mind the following:

- To minimize complexity, try to use only one management strategy per Deployment.
- Your strategy for managing connections and variables locally should be compatible with your management strategy locally. For example, if you use a secrets backend on Astro, you might want to use test data from the same secrets backend to test your DAGs locally. Read the following topics to learn more about each available strategy for local development. 

### Manage connections and variables in the Airflow UI for your Deployment

Similarly to in your local environment, you can create connections and variables in the Airflow UI on your Deployment. This strategy is good if you're just getting started with Airflow or you're the only person with access to your Deployment.

The limitations of this strategy are:

> - Any Workspace Admin can view your connections and variables. 

Reviewer 4: not sure if this is a limitation, I would rather argue it is a feature of Astro

> - Managing many connections or variables this way can become unwieldy.

Reviewer 4: Connections can be tested as well from the UI easily.

- It's not the most secure option for sensitive variables. 

> - You can't pull your variables down into a local project. 

Reviewer 4: this is not true, we can export the Airflow variables from Airflow and import to any other Airflow.

To create Airflow connections and variables in the Airflow UI for your Deployment:

1. In the Cloud UI, select a Workspace, then select a Deployment.
2. Click **Open Airflow**. 
3. Create your connections and variables. See:

    - [Managing Variables](https://airflow.apache.org/docs/apache-airflow/stable/howto/variable.html)
    - [Create a Connection in the UI](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#creating-a-connection-with-the-ui)

### Store connections and variables as environment variables

> You can use Astro [environment variables](environment-variables.md) to store Airflow connections and variables. This strategy is good if you don't have a secrets backend, but you still want to take advantage of security and RBAC features to limit access to connections and variables.

Reviewer 3: There's a lot of spots in this doc where we make claims like "this is good", "this is the easiest" - how do we feel about that kind of language? I like docs that provide opinions and recommendations, but not sure about the best way to compose them.

The benefits of managing Airflow connections and variables in your `.env` file are: 

- You can pull environment variables down to your local computer using the Astro CLI, making it easy to import Airflow resources from Astro.
- Your secret values are encrypted on the Astronomer control plane.

The limitations of this method are:

- Connections can't be viewed or managed from the Airflow UI.
- Connections are more difficult to format as URIs or JSON.
- It's not as secure or centralized as a secrets backend.

#### Set Airflow connections as Astro environment variables

Create an Astro environment variable to store your connection. See [Set environment variables in the Cloud UI](https://docs.astronomer.io/astro/environment-variables#set-environment-variables-in-the-cloud-ui).

In Astro Runtime version 4.2 and earlier, use the Airflow [connection URI format](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#uri-format) to store connections as environment variables. The naming convention for Airflow environment variable connections is:

- Key: `AIRFLOW_CONN_<CONN_ID>` 
- Value: `<connection-uri>`

> In Astro Runtime version 5.0 and later, you can also use JSON format to store connections. See [JSON format example](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#json-format-example). When using the JSON format in environment variables, the JSON object must be defined in a single, unbroken line. 

Reviewer 3: ```suggestion
In Astro Runtime version 5.0 and later, you can also use JSON format to store connections or the Connection URI format. See [JSON format example](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#json-format-example). When using the JSON format in environment variables, the JSON object must be defined in a single, unbroken line. 
```

What's the "also" referring to?

#### Set Airflow variables as Astro environment variables

Create an Astro environment variable to store your Airflow variable. See [Set environment variables in the Cloud UI](https://docs.astronomer.io/astro/environment-variables#set-environment-variables-in-the-cloud-ui).

The environment variable naming convention for Airflow variables is:

- Key: `AIRFLOW_VAR_<VAR_NAME>` 
- Value: `<value>`

#### Import environment variables to a local project

If you are storing your connections and variables in Astro environment variables, run the following command to pull the keys and non-secret values of your environment variables to a local `.env` file:

```bash
astro deployment variable list --deployment-name "<deployment name>" --save --env <env-file-location>
```

> This command does not pull connections or variables that were set in the Airflow UI, nor does it pull down any environment variable values stored as secrets. 

Reviewer 3: ```suggestion
This command does not pull connections or variables that were set in the Airflow UI. It also does not pull any environment variable values stored as secrets. 
```
